README.md
# pydvdid-m

![downloads](https://pepy.tech/badge/pydvdid-m)
![license](https://img.shields.io/pypi/l/pydvdid-m.svg)
![wheel](https://img.shields.io/pypi/wheel/pydvdid-m.svg)
![versions](https://img.shields.io/pypi/pyversions/pydvdid-m.svg)

Pure Python implementation of the Windows API method [IDvdInfo2::GetDiscID].  
This is a modification of [sjwood's pydvdid](https://github.com/sjwood/pydvdid).

The Windows API method [IDvdInfo2::GetDiscID] is used by Windows Media Center to compute a
'practically unique' 64-bit CRC for metadata retrieval. It's metadata retrieval API has
sadly since shutdown around October 2019 and all it's information is presumably lost.

  [IDvdInfo2::GetDiscID]: <https://docs.microsoft.com/en-us/windows/win32/api/strmif/nf-strmif-idvdinfo2-getdiscid>

## Changes compared to sjwood's repo

1. License changed from Apache-2.0 to GPL-3.0.
2. Moved build tools and dependency management from setuptools and requirements.txt to poetry.
3. Support for Python 2.x and Python <3.6 has been dropped. 
4. All tests were removed entirely simply because a lot of the tests would need to be refactored
   for general code changes, and some tests might not be needed anymore.
5. All custom exceptions were removed entirely and replaced with built-in ones.
6. CRC-64 related code were refactored and merged as one CRC64 class in one file.
7. The merged CRC64 class contains various improvements over the original code, including
   improvements with doc-strings, formatting, and such.
8. Various BASH shell scripts and config files were removed entirely as they are deemed unnecessary.
9. Uses pycdlib to read from ISO and direct disc drives, instead of assuming that it's a folder.

Other than that, the rest of the changes are general code improvements in various ways.
There may be more differences as the repo gets commits, but these are the primary differences from
[sjwood's commit](https://github.com/sjwood/pydvdid/commit/03914fb7e24283c445e5af724f9d919b23caaf95) to
the beginnings of this repository.

## Important Information on DVD ID Accuracy

1. The DVD ID generated assumes that the input Disc, ISO, or VIDEO_TS folder has the original untouched
   file timestamps, file names, and header data. Any change or missing file will result in a different DVD ID.
2. Because of this, AnyDVD HD, DVDFab Passkey, or anything similar that may change the disc data at a
   driver-level should be disabled before the use of pydvdid-m.
3. Just because it is an ISO file, does not mean it is truly untouched in the header areas that matter.
   AnyDVD HD (and maybe others) may alter some bytes here and there when removing protection.
4. VIDEO_TS folders are typically created by extracting the data from an ISO or Disc, which will most likely
   re-generate file creation and modified timestamps. If you want truly accurate DVD IDs, I cannot advise
   the use of this project on VIDEO_TS folders.

If you want truly accurate DVD IDs, then only use this project direct from discs with all DVD ripping software
disabled and closed. Make sure nothing like AnyDVD HD or DVDFab Passkey is running in the background.

## Installation

```shell
$ pip install pydvdid-m
```

## Usage

You can run DvdId on all types of DVD video file structures:

- Direct from Disc, e.g., `/dev/sr0`, `\\.\E:`, or such.
- An ISO file, e.g., `/mnt/cdrom`, `C:/Users/John/Videos/FAMILY_GUY_VOLUME_11_DISC_1.ISO`.
- A VIDEO_TS folder*, `C:/Users/John/Videos/THE_IT_CROWD_D1/`.

Note: Generating a DVD ID from a VIDEO_TS folder has a high chance of providing an
invalid DVD ID. The algorithm uses file creation timestamps, and extracting VIDEO_TS folders
direct from Disc or from an ISO will most likely change them, especially when transferred or moved.

### CLI

```shell
phoenix@home@~$ dvdid "FAMILY_GUY_VOLUME_11_DISC_1.ISO"
<Disc>
<Name>FAMILY_GUY_VOLUME_11_DISC_1</Name>
<ID>db3804e3|1645f594</ID>
</Disc>
```

You can provide a path to an ISO file, or a mounted device, e.g.:

```shell
phoenix@home@~$ dvdid "/dev/sr0"
<Disc>
<Name>BBCDVD3508</Name>
<ID>3f041dfc|27ffd3a8</ID>
</Disc>
```

or on Windows via Raw Mounted Device:

```shell
PS> dvdid "\\.\E:"
<Disc>
<Name>BBCDVD3508</Name>
<ID>3f041dfc|27ffd3a8</ID>
</Disc>
```

### Package

You can also use pydvdid-m in your own Python code by importing it.  
Here's a couple of things you can do, and remember, you can use both ISO paths and mounted device targets.

```python
>>> from pydvdid_m import DvdId
>>> dvd_id = DvdId(r"C:\Users\John\Videos\FAMILY_GUY_VOLUME_11_DISC_1.ISO")
>>> dvd_id.disc_label
'BBCDVD3508'
>>> repr(dvd_id.checksum)
'<CRC64 polynomial=0x92c64265d32139a4 xor=0xffffffffffffffff checksum=0x3f041dfc27ffd3a8>'
>>> dvd_id.checksum
'3f041dfc|27ffd3a8'
>>> dvd_id.checksum.as_bytes
b"?\x04\x1d\xfc'\xff\xd3\xa8"
>>> dvd_id.dumps()
'<Disc>\n<Name>BBCDVD3508</Name>\n<ID>3f041dfc|27ffd3a8</ID>\n</Disc>'
```

## License

[GNU General Public License, Version 3](https://raw.githubusercontent.com/rlaphoenix/pydvdid-m/master/LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.1] - 2022-01-20

### Added

- Added important information on DVD ID accuracy to the README.
- Added a "Saved DVD ID to" print statement in the primary script to show it got auto-saved, and to where.

### Changed

- Ensured that filenames are uppercase. ISO-9660 isn't strict about file-casing, but the DVD spec is.
- Changed `DvdId.dump`'s return value from the `Path.write_text` int return value to the Path which
  was written to.

### Fixed

- Fixed raw Windows Device target checks, e.g. `\\.\E:` from being detected as a VIDEO_TS folder.

## [1.1.0] - 2022-01-15

### Added 

- This CHANGELOG file.
- Support for VIDEO_TS folders. However, they may not provide an accurate DvdId, see README.
- Added support for Path targets in disc_label. It returns the folder/file name as the disc label.
- Added file extension restrictions for files that get processed in DvdId (BUP, IFO, VOB).
- Added function `DvdId._get_file` to get a direct path (as the correct object type) to a specific file.

### Changed

- Refactored DvdId's class variable `disc_label` as a function property.
- Ensured that `VIDEO_TS.IFO` would be processed before `VTS_01_0.IFO`, and that they would both be processed.

### Fixed

- Fixed mistake in `DvdId._get_first_64k_content` which had the variables of the expected/read bytes mixed up.
- Fixed possible invalid creation time seconds value if it was somehow in floating-point accuracy. DVD IDs made
  from ISO files or straight from disc shouldn't have had any issues.
- Corrected the Type-hint of `UDFFileEntry` to `DirectoryRecord`.
- Added Error Handling to `DvdId._get_files`, which could cause an exception if the path isn't found.
- Fixed the `/VIDEO_TS` directory exists check in DvdId.

## [1.0.0] - 2022-01-15

### Added

- CRC64 class with various formatting helper functions within it.
- DvdId class that can generate a DVD ID for ISO files and direct from disc.
- Executable script via poetry that executes `DvdId.dump()` and `DvdId.dumps()` on the input.
- GitHub Workflows for CI/CD.
- Poetry Dependency Management and Building tooling.

[unreleased]: https://github.com/rlaphoenix/pydvdid-m/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/rlaphoenix/pydvdid-m/releases/tag/v1.0.0


pydvdid_m/crc64.py
from __future__ import annotations

import struct
from typing import Union


class CRC64:
    def __init__(self, polynomial: int, initial_xor: int = 0xffffffffffffffff):
        """
        Calculates a 64-bit Cyclic Redundancy Check checksum.

        Requires a polynomial to seed the construction of a lookup table and an
        optional initial XOR value.
        """
        if not isinstance(polynomial, int):
            raise ValueError(f"Polynomial must be an integer, not {polynomial!r}.")
        if not isinstance(initial_xor, int):
            raise ValueError(f"Initial XOR must be an integer, not {initial_xor!r}.")

        self._polynomial = polynomial
        self._initial_xor = initial_xor

        self._lookup_table = self._construct_lookup_table(polynomial)
        self._checksum = initial_xor

    def __eq__(self, other: CRC64) -> bool:
        return self._checksum == other._checksum

    def __ne__(self, other: CRC64) -> bool:
        return not self.__eq__(other)

    def __str__(self) -> str:
        return f"{self.high_bytes}|{self.low_bytes}"

    def __repr__(self) -> str:
        return f"<CRC64 " \
               f"polynomial=0x{self._polynomial:016x} " \
               f"xor=0x{self._initial_xor:016x} " \
               f"checksum=0x{self._checksum:016x}" \
               ">"

    @property
    def high_bytes(self) -> str:
        """Topmost 4 bytes of the checksum formatted as a lowercase hex string."""
        return format(self._checksum >> 32, "08x")

    @property
    def low_bytes(self) -> str:
        """Bottommost 4 bytes of the checksum formatted as a lowercase hex string."""
        return format(self._checksum & 0xffffffff, "08x")

    @property
    def as_bytes(self) -> bytes:
        """Checksum represented as bytes."""
        return struct.pack(">Q", self._checksum)

    def update(self, data: Union[bytes, bytearray]) -> None:
        """Enumerate the bytes of the supplied bytearray and update the CRC-64 in-place."""
        if not isinstance(data, (bytes, bytearray)):
            raise ValueError(f"Data must be a byte array, not {data!r}.")

        for byte in data:
            self._checksum = (self._checksum >> 8) ^ self._lookup_table[(self._checksum & 0xff) ^ byte]

    @staticmethod
    def _construct_lookup_table(polynomial: int) -> list[int]:
        """Precomputes the CRC-64 lookup table seeded from the supplied polynomial."""
        if not isinstance(polynomial, int):
            raise ValueError(f"Polynomial must be an integer, not {polynomial!r}.")

        lookup_table = []
        for i in range(0, 256):
            lookup_value = i
            for _ in range(0, 8):
                if lookup_value & 0x1 == 0x1:
                    lookup_value = (lookup_value >> 1) ^ polynomial
                else:
                    lookup_value = lookup_value >> 1
            lookup_table.append(lookup_value)
        return lookup_table


pydvdid_m/dvdid.py
from datetime import datetime
from inspect import cleandoc
from io import BytesIO
from pathlib import Path
from struct import pack_into
from typing import Union, Iterator, Optional

from dateutil.tz import tzoffset
from pycdlib import PyCdlib
from pycdlib.dr import DirectoryRecord
from pycdlib.pycdlibexception import PyCdlibInvalidInput

from pydvdid_m.crc64 import CRC64


class DvdId:
    def __init__(self, target: Union[str, PyCdlib], allow_folder_id: bool = False):
        """
        Computes a Windows API IDvdInfo2::GetDiscID-compatible 64-bit Cyclic Redundancy Check
        checksum from the DVD .vob, .ifo, and .bup files found in the supplied DVD device.
        """
        if isinstance(target, str):
            target_path = Path(target)
            if target_path.is_dir() and not target.startswith("\\\\.\\"):
                if allow_folder_id or input(
                    "Warning: Extracted VIDEO_TS folders most likely have modified file timestamps. "
                    "You may receive an inaccurate DVD ID if the timestamp does not match whats stated "
                    "in the original ISO 9660 headers.\n\n"
                    "Do you wish to continue anyway? (y/N): "
                ).lower() != "y":
                    return
                self.device = target_path
            else:
                # assume path to an ISO
                self.device = PyCdlib()
                self.device.open(target)
        elif isinstance(target, PyCdlib):
            self.device = target
        else:
            raise ValueError(f"Unsupported target: {target}")

        if not any(self._get_files("/VIDEO_TS")):
            # Not a DVD-Video Disc?
            raise FileNotFoundError(f"The /VIDEO_TS directory and it's files doesn't exist in {target}")

        # the polynomial used for this CRC-64 checksum is:
        # x^63 + x^60 + x^57 + x^55 + x^54 + x^50 + x^49 + x^46 + x^41 + x^38 + x^37 + x^34 + x^32 +
        # x^31 + x^30 + x^28 + x^25 + x^24 + x^21 + x^16 + x^13 + x^12 + x^11 + x^08 + x^07 + x^05 + x^2
        crc = CRC64(0x92c64265d32139a4)

        for file in self._get_files("/VIDEO_TS"):
            if self._get_file_name(file, as_string=True).upper().split(".")[-1] in ("BUP", "IFO", "VOB"):
                crc.update(self._get_file_creation_time(file))
                crc.update(self._get_file_size(file))
                crc.update(self._get_file_name(file))

        crc.update(self._get_first_64k_content(self._get_file("/VIDEO_TS/VIDEO_TS.IFO")))
        crc.update(self._get_first_64k_content(self._get_file("/VIDEO_TS/VTS_01_0.IFO")))

        self.checksum = crc

    @property
    def disc_label(self) -> Optional[str]:
        """Get Disc Label. Returns the directory Name if not an ISO or Device."""
        if isinstance(self.device, PyCdlib):
            return self.device.pvd.volume_identifier.replace(b"\x00", b"").strip().decode()
        elif isinstance(self.device, Path):
            return self.device.name
        else:
            return None

    def dumps(self) -> str:
        """Return DVD ID with Disc Label in XML format."""
        xml = cleandoc("""
        <Disc>
        <Name>%s</Name>
        <ID>%s</ID>
        </Disc>
        """)
        if self.disc_label:
            if not isinstance(self.disc_label, str):
                raise ValueError("disc_label must be a string")
        xml = xml % (self.disc_label or "", str(self.checksum))
        return xml

    def dump(self, to: Union[Path, str]) -> Path:
        """
        Save DVD ID with Disc Label in XML format.

        Path can be to a direct file path or directory path.
        A custom filename using the Disc Label will be used for you
        if it's a path to a directory or non-XML file.

        Returns the Path where the DVD ID was saved.
        """
        if not to:
            raise ValueError("A save path must be provided.")
        if not isinstance(to, Path):
            to = Path(to)
        if to.is_dir():
            to = to / f"{self.disc_label}.dvdid.xml"
        if not to.suffix.lower().endswith(".xml"):
            to = to.with_suffix(f".{self.disc_label}.dvdid.xml")
        to.parent.mkdir(parents=True, exist_ok=True)
        to.write_text(self.dumps(), encoding="utf8")
        return to

    def _get_files(self, iso_path: str) -> Iterator[DirectoryRecord]:
        """
        Yield all files in the device at the provided ISO path.
        Note: Special paths `.` and `..` paths are not yielded.
        """
        if isinstance(self.device, PyCdlib):
            try:
                for dr in self.device.list_children(iso_path=iso_path):
                    if dr.file_identifier().decode() in (".", ".."):
                        continue
                    yield dr
            except PyCdlibInvalidInput:
                pass  # path probably doesnt exist
        elif isinstance(self.device, Path):
            for file in (self.device / iso_path.lstrip("\\/")).iterdir():
                yield file
        else:
            raise ValueError(f"Target {self.device!r} unsupported.")

    def _get_file(self, iso_path: str) -> Union[DirectoryRecord, Path]:
        path = Path(iso_path.lstrip("\\/"))
        for file in self._get_files(f"/{path.parent}"):
            if self._get_file_name(file, as_string=True).upper() == path.name.upper():
                return file
        raise FileNotFoundError(f"File {path} could not be found.")

    def _get_first_64k_content(self, file: Union[DirectoryRecord, Path]) -> bytes:
        """
        Returns the first 65536 (or the file size, whichever is smaller) bytes of the file at the
        specified file path, as a bytearray.
        """
        if isinstance(file, DirectoryRecord):
            file_size = file.get_data_length()
        else:
            file_size = file.stat().st_size
        read_size = min(file_size, 0x10000)

        if isinstance(file, DirectoryRecord):
            f = BytesIO()
            self.device.get_file_from_iso_fp(
                outfp=f,
                iso_path=f"/VIDEO_TS/{file.file_identifier().decode()}"
            )
            f.seek(0)
            content = f.read(read_size)
        else:
            content = bytearray(read_size)
            with file.open("rb") as f:
                f.readinto(content)  # type: ignore

        if content is None or len(content) < read_size:
            raise EOFError(f"{read_size} bytes were expected, {len(content or [])} were read.")

        return content

    @staticmethod
    def _get_file_name(file: Union[Path, DirectoryRecord], as_string: bool = False) -> Union[bytearray, str]:
        """Get the name of the Directory Record as a UTF-8 bytearray terminated with a NUL character."""
        if isinstance(file, DirectoryRecord):
            file_identifier = file.file_identifier().decode()
            file_name = file_identifier.split(";")[0]  # remove ;N (file version)
        else:
            file_name = file.name
        file_name = file_name.upper()  # linux may allow it as other casing
        if as_string:
            return file_name
        utf8_name = bytearray(file_name, "utf8")
        utf8_name.append(0)
        return utf8_name

    @staticmethod
    def _get_file_size(file: Union[Path, DirectoryRecord]) -> bytearray:
        """Get the size of the Directory Record formatted as a 4-byte unsigned integer bytearray."""
        if isinstance(file, DirectoryRecord):
            file_size = file.get_data_length()
        else:
            file_size = file.stat().st_size
        array = bytearray(4)
        pack_into(b"I", array, 0, file_size)
        return array

    @staticmethod
    def _get_file_creation_time(file: Union[Path, DirectoryRecord]) -> bytearray:
        """
        Get the creation time in Microsoft FILETIME structure as a 8-byte unsigned integer bytearray.
        https://msdn.microsoft.com/en-us/library/windows/desktop/ms724284.aspx
        """
        if isinstance(file, DirectoryRecord):
            dr_date = file.date
            dr_date = datetime(
                year=1900 + dr_date.years_since_1900, month=dr_date.month, day=dr_date.day_of_month,
                hour=dr_date.hour, minute=dr_date.minute, second=dr_date.second,
                # offset the timezone, since ISO's dates are offsets of GMT in 15 minute intervals, we
                # need to calculate that but in seconds to pass to tzoffset.
                tzinfo=tzoffset("GMT", (15 * dr_date.gmtoffset) * 60)
            )
            epoch_offset = dr_date - datetime(1601, 1, 1, tzinfo=tzoffset(None, 0))
        else:
            # TODO: The created time may have been modified or even just differ after
            #       extraction, any difference to the true timestamp will return a
            #       different DVD ID. This is a problem.
            c_time = file.stat().st_ctime
            if c_time < -11644473600 or c_time >= 253402300800:
                raise ValueError(f"Created timestamp for file {file.name} is out of range: {c_time}")
            epoch_offset = datetime.utcfromtimestamp(c_time) - datetime(1601, 1, 1)

        filetime = int(int(epoch_offset.total_seconds()) * (10 ** 7))
        file_creation_time = bytearray(8)
        pack_into(b"Q", file_creation_time, 0, filetime)

        return file_creation_time
